CS483/CS483/Kartaclysm/Services/IO/PlayerInputMapping.cpp

```diff
@@ -0,0 +1,334 @@
+	// Note: Make sure you pass a player number less than what is set by SetSplitscreenPlayers().
+	//--------------------------------------------------------------------------------
+	bool PlayerInputMapping::ManuallyMapPlayer(const int p_iPlayer, const int p_iGLFWJoystick)
+	{
+		assert(p_iPlayer <= m_iPlayersRacing);
+
+		if (p_iGLFWJoystick == -1) // disable
+		{
+			(*m_pPlayers)[p_iPlayer] = -1;
+			return true;
+		}
+
+		// Check if desired input is in use
+		PlayerMap::iterator it = m_pPlayers->begin(), end = m_pPlayers->end();
+		bool bFound = false;
+		for (; it != end; it++)
+		{
+			if (it->second == p_iGLFWJoystick)
+			{
+				bFound = true;
+				break;
+			}
+		}
+
+		if (!bFound) // input not in use: assign it
+		{
+			(*m_pPlayers)[p_iPlayer] = it->second;
+			return true;
+		}
+
+		return false;
+	}
+
+	//--------------------------------------------------------------------------------
+	// PlayerInputMapping::GetFirstAvailableInput
+	// Returns: int
+	//
+	// Returns Joystick number, or GLFW_JOYSTICK_LAST + 1 for keyboard, or -1 if none available
+	//--------------------------------------------------------------------------------
+	int PlayerInputMapping::GetFirstAvailableInput()
+	{
+		for (int i = 0; i <= GLFW_JOYSTICK_LAST + 1; i++)
+		{
+			if (i == GLFW_JOYSTICK_LAST + 1 || glfwJoystickPresent(i))
+			{
+				// See if input is in use
+				bool bFound = false;
+				PlayerMap::iterator it = m_pPlayers->begin(), end = m_pPlayers->end();
+				for (; it != end; it++)
+				{
+					if (it->second == i)
+					{
+						bFound = true;
+						break;
+					}
+				}
+
+				// If not in use, return it
+				if (!bFound)
+				{
+					return i;
+				}
+			}
+		}
+
+		// No available inputs
+		return -1;
+	}
+
+	//--------------------------------------------------------------------------------
+	// PlayerInputMapping::Update
+	// Parameter: float fDelta
+	//
+	// Iterate through players and call method for their input types
+	//--------------------------------------------------------------------------------
+	void PlayerInputMapping::Update(const float p_fDelta)
+	{
+		PlayerMap::const_iterator it = m_pPlayers->begin(), end = m_pPlayers->end();
+		for (; it != end; it++)
+		{
+			if (it->second != -1) // keyboard or joystick connected for this player
+			{
+				InputActionMapping::Instance()->SendInputEventForPlayer(it->first, it->second);
+			}
+			else
+			{
+				// If a player is not connected, you should not be calling this during raceMode
+				assert(m_bRaceMode);
+			}
+		}
+	}
+
+	//--------------------------------------------------------------------------------
+	// PlayerInputMapping::JoystickCallback
+	// Parameter:	const HeatStroke::Event* p_pEvent - "JoystickCallback" event
+	//
+	// Adds or removes joystick from player input mapping.
+	//--------------------------------------------------------------------------------
+	void PlayerInputMapping::JoystickCallback(const HeatStroke::Event* p_pEvent)
```

Comment:
Agreed, it is rather hacky. I fully plan to come back to it when we do split screen, but for now it does its job of switching between gamepad and keyboard automatically.